### ISSUE TITLE
Without making change to our CI, all environments created from cache now failing for any commit.  

### ISSUE THREAD

**Issue by priti-ashvin-shah-ibm** [tox-dev/tox#2771]
## Issue

Without making change to our CI, all environments created using tox were failing for every commit.  After we limited the version of tox 3.28.0, then we can pass CI.  This is a temporary solution.  We would appreciate some advice so we don't have to limit tox versions within our CI.  Our repo is open source and at https://github.com/Qiskit/qiskit-metal

## Environment
This is happening in github actions, not locally.   https://github.com/Qiskit/qiskit-metal/blob/main/.github/workflows/main.yml

Provide at least:

- OS:
- `pip list` of the host Python where `tox` is installed:
We tried on Ubuntu 20.04 passed, mac-os -latest passed
```console
Not on command line, but in github actions. 
```

## Output of running tox

Provide the output of `tox -rvv`:

```console

```

## Minimal example

If possible, provide a minimal reproducer for the issue:
We can't run github actions locally, so sorry that we can't provide this. However, we would appreciate your support. 

```console

```

**Comment by priti-ashvin-shah-ibm** [comment]
@obrienpja

**Comment by jugmac00** [comment]
@priti-ashvin-shah-ibm at very least we need a link to a failing build

**Comment by priti-ashvin-shah-ibm** [comment]
Sure, https://github.com/Qiskit/qiskit-metal/commits/main
The builds between 
<img width="1285" alt="image" src="https://user-images.githubusercontent.com/74020801/209208954-0b582f38-a463-4fae-8ae2-38bbc427e639.png">


Nicks iswap had passed CI, then inadvertently waited to merge, after the merge, the CI failed.

**Comment by priti-ashvin-shah-ibm** [comment]
I am happy to provide anything more that would be helpful.

**Comment by gaborbernat** [comment]
Please fill out the parts in the template.

**Comment by priti-ashvin-shah-ibm** [comment]
This is happening in github actions, not locally. https://github.com/Qiskit/qiskit-metal/blob/main/.github/workflows/main.yml

**Comment by gaborbernat** [comment]
You can fill those sections then by entering the data from GitHub action runs.

**Comment by priti-ashvin-shah-ibm** [comment]
This is a snippet from an issue from github actions setup-python.

https://github.com/actions/setup-python/issues/570

You install tox in scope of the [Install Deps step](https://github.com/Qiskit/qiskit-metal/blob/main/.github/workflows/main.yml#L39). Tox 3.27.1 was installed during the [last successful build](https://github.com/Qiskit/qiskit-metal/actions/runs/3630793091/jobs/6124579484#step:5:10) and tox 4.0.8 was installed during the [first failed build](https://github.com/Qiskit/qiskit-metal/actions/runs/3686915714/jobs/6239805065#step:5:10). It looks like your tests don't work with the new major version of tox.

To fix the issue, please adjust your tests or install the previous major version of tox:

      - name: Install Deps
        run: |
          python -m pip install -U git+https://github.com/tox-dev/tox.git@3.28.0 setuptools virtualenv wheel
          sudo apt install libglu1-mesa libglu1-mesa-dev

**Comment by priti-ashvin-shah-ibm** [comment]
> 
Sure, Will gather that.  Thank you for spending the time to help us.  Just FYI: This is an open-source repo that anyone can look at.

**Comment by priti-ashvin-shah-ibm** [comment]
The below image is a commit that failed. The interesting part is that the same commit passed before the merge. When the merge happens, CI is run again.  So the same exact commit failed with CI after two weeks. The time between passing and failing during the merge was about 2 weeks. Also, nothing in github actions was changed.

![image](https://user-images.githubusercontent.com/74020801/209222611-47c99422-17aa-460c-ac47-a5be41aab037.png)


This is the same commit just before the merge where it passed.

![image](https://user-images.githubusercontent.com/74020801/209222783-5719c2f2-f1ab-49af-8756-c0cae3fc95b1.png)

**Comment by q0w** [comment]
@priti-ashvin-shah-ibm I think, you have the same issue https://github.com/tox-dev/tox/issues/2730
tldr: you should disable global skipsdist and use skip_install=True on your docs/lint actions
```diff
diff --git a/tox.ini b/tox.ini
index e258456..74d1677 100644
--- a/tox.ini
+++ b/tox.ini
@@ -1,7 +1,6 @@
 [tox]
 minversion = 2.1
 envlist = py39, lint, docs
-skipsdist = True
 
 [testenv]
 usedevelop = true
@@ -17,6 +16,7 @@ commands =
 
 [testenv:lint]
 basepython = python3.9
+skip_install = true
 deps =
   pylint
   ipython
@@ -27,6 +27,7 @@ commands =
 
 [testenv:docs]
 basepython = python3.9
+skip_install = true
 envdir = .tox/docs
 passenv = DOCS_FROM_MASTER QISKIT_DOCS_BUILD_TUTORIALS
 deps =
@@ -38,6 +39,7 @@ commands =
 [testenv:gettext]
 basepython = python3.9
 envdir = .tox/docs
+skip_install = true
 deps =
   -r requirements-dev.txt
   sphinx-intl

**Comment by jugmac00** [comment]
Thanks @q0w - closed as duplicate of #2730

**Comment by priti-ashvin-shah-ibm** [comment]
Thank you @q0w for your time to share a proposed fix.  I made a branch and pull request on the Qiskit-Metal repository.  I made a draft pull request to force CI.  The code change included your fix along with using the latest version of tox. Unfortunately, there is still an error with tox for edocs.  Any idea as to where to fix this?  I would appreciate your insight. 
https://github.com/Qiskit/qiskit-metal/pull/893

<img width="1668" alt="image" src="https://user-images.githubusercontent.com/74020801/209285534-6900b740-4ac7-4a55-9ea4-e5e59adc455b.png">

**Comment by priti-ashvin-shah-ibm** [comment]
@jugmac00 , please keep this open until we are able to use the latest version of tox on our open-source repository at Qiskit-Metal.

**Comment by gaborbernat** [comment]
The error is fairly self explanatory.

**Comment by jugmac00** [comment]
@priti-ashvin-shah-ibm The error message says that you need to use a comma instead of a space as a separator, also see https://tox.wiki/en/latest/faq.html#tox-4-changed-ini-rules

**Comment by priti-ashvin-shah-ibm** [comment]
Yes, thank you.  The person who wrote the tox.ini has left our group.  Wish tox was backward compatible.  Now there is another bug.

**Comment by jugmac00** [comment]
Hey @priti-ashvin-shah-ibm I am sorry that you face some issues with configuring `tox`.

tl/dr I have left a comment which should fix your issue on your pull request.

> Wish tox was backward compatible.

In an ideal world it would be - we had to make some breaking changes to streamline tox, and also there were some usages of the previous version of tox which never should have worked, but worked by chance. The latest release was a major new version, a complete rewrite, it is not expected that we will do that any time soon again :-)

I think you should be good to go now. If you have any further issues, please do not hesitate to ask again.

> Now there is another bug. CI is still not passing.

That is not a bug - you just used the wrong syntax. The correct syntax is in our [documentation](https://tox.wiki/en/latest/config.html#allowlist_externals), which you maybe did not have the chance to read. While I do understand your frustration, at one point it would make sense to sit down and read the documentation of a tool you want to use.

**Comment by priti-ashvin-shah-ibm** [comment]
Really do appreciate your help.  However I did try the syntax that you proposed.  That also failed.

**Comment by priti-ashvin-shah-ibm** [comment]
Can we leave this issue open until next week?  I don't have the bandwidth for this today.

**Comment by jugmac00** [comment]
> However I did try the syntax that you proposed. That also failed.

You did not - you missed an equal sign.

> Can we leave this issue open until next week? I don't have the bandwidth for this today.

There is no `tox` issue any more. You just had a typo / copy/paste error.

When things go wrong, they go really wrong :-) Happens to all of us.

I won't be around the next days - I celebrate xmas with my family.

I do hope you will be able to fix the issues yourself - if not, just give us a ping and we'll have a look (after xmas).

**Comment by priti-ashvin-shah-ibm** [comment]
Thank you @jugmac00 !!

**Comment by priti-ashvin-shah-ibm** [comment]
@jugmac00 , We made some progress and are able to use tox v4 on main.  However, within tox.ini on our repo. we had to make skip_install=False for testenv:docs.  This allowed the environment to be made with requirements.txt along with requirement-dev.txt.  Can you share if there is  tox syntax to **explicitly** use both files?

**Comment by priti-ashvin-shah-ibm** [comment]
We can close this issue for now.

**Comment by jugmac00** [comment]
Hey @priti-ashvin-shah-ibm

I already tried to explain the issue a couple of days ago in https://github.com/Qiskit/qiskit-metal/pull/893#discussion_r1056427624 - did you see the comment? Is there anything unclear?

Also, when I have a look at https://github.com/Qiskit/qiskit-metal/blob/9722b3bd62a5d63f8f8b1607ce81397f12e8d41b/tox.ini#L28-L37 - this contradicts your above comment.

> Can you share if there is tox syntax to explicitly use both files?

tox explicitly installs things via `deps`, that is what you did. And tox implicitly builds and installs a package e.g. via the information in `setup.py` unless you add a `skip_install = true`.

I think a better part of the confusion was created by your package using `requirements.txt` as `install_requires` in your `setup.py`.

Both `requirements.txt` and `setup.py` serve different purposes and should not be mixed.

**Comment by priti-ashvin-shah-ibm** [comment]
Hi @jugmac00 , 
I had meant to post here after [this](url) merge to give an update.  I believe we have environments using the latest version of tox and ubuntu and commits are passing as expected now.  Thank you for you time and support for helping us.
